frontend/javascripts/oxalis/model/bucket_data_handling/data_cube.js

```diff
@@ -367,6 +377,120 @@ class DataCube {
+      const [currentBucket, initialAddress] = bucketsToFill.pop();
+      // Check if the buckets overlaps the active viewport bounds.
+      if (!areBoundingBoxesOverlapping(currentBucket.getBoundingBox(), viewportBoundings)) {
+        continue;
+      }
+      const bucketData = currentBucket.getOrCreateData();
+      const currentVoxel = this.getVoxelIndex(initialAddress, zoomStep);
+      if (bucketData[currentVoxel] !== sourceCellId) {
+        // Ignoring neighbour buckets whose cellId at the initial position does not match the source cell id.
+        continue;
+      }
+      // Add the bucket to the current volume undo batch, if it isn't already part of it.
+      const zoomedAddressAsString = currentBucket.zoomedAddress.toString();
+      if (!bucketsAlreadyInUndoState.has(zoomedAddressAsString)) {
+        bucketsAlreadyInUndoState.add(zoomedAddressAsString);
+        Store.dispatch(
+          addBucketToUndoAction(currentBucket.zoomedAddress, currentBucket.getCopyOfData()),
+        );
+      }
+      // Use a Dynamic2DVoxelIterator to iterate over the bucket in 2d and using bucket-local addresses and not global addresses.
+      const neighbourVoxelIterator = new Dynamic2DVoxelIterator(get2DAddress(initialAddress));
+      let labeledVoxel = 0;
+      // Iterating over all neighbours from the initialAddress.
+      while (neighbourVoxelIterator.notEmpty()) {
+        const neighbours = neighbourVoxelIterator.getNeighbors();
+        for (let neighbourIndex = 0; neighbourIndex < neighbours.length; ++neighbourIndex) {
+          const neighbourVoxel = neighbours[neighbourIndex];
+          let neighbourVoxel3D = get3DAddress(neighbourVoxel, initialVoxel);
+
+          // If the current neighbour is not in the current bucket, calculate its
+          // bucket's zoomed address and add the bucket to bucketsToFill.
+          let neighbourBucketAddress: ?Vector4 = null;
+          for (let dimensionIndex = 0; dimensionIndex < 2; ++dimensionIndex) {
+            const dimension = dimensionsToIterateOver[dimensionIndex];
+            if (neighbourVoxel[dimensionIndex] < 0) {
+              neighbourBucketAddress = [
+                currentBucket.zoomedAddress[0],
+                currentBucket.zoomedAddress[1],
+                currentBucket.zoomedAddress[2],
+                zoomStep,
+              ];
+              neighbourBucketAddress[dimension] -= 1;
+              // Add a full bucket width to the coordinate below 0 to avoid error's caused by the modulo operation.
+              neighbourVoxel[dimensionIndex] += constants.BUCKET_WIDTH;
+            } else if (neighbourVoxel[dimensionIndex] >= constants.BUCKET_WIDTH) {
+              neighbourBucketAddress = [
+                currentBucket.zoomedAddress[0],
+                currentBucket.zoomedAddress[1],
+                currentBucket.zoomedAddress[2],
+                zoomStep,
+              ];
+              neighbourBucketAddress[dimension] += 1;
```

Comment:
Also, it could make sense to add comments to some of the variable names, as these are not always self-explanatory from my POV :)

Reply:
> Also, it could make sense to add comments to some of the variable names, as these are not always self-explanatory from my POV :)

I need some clarification here which variables are not self-explanatory to you.

Reply:
Nevermind, not sure, for which names I had questions. Probably not relevant anymore  :)